File: benchmarks/pandas/bench_dropna_fn.py
"""
Benchmark: pandas Series.dropna() / DataFrame.dropna() — drop missing values.
Outputs JSON: {"function": "dropna_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import pandas as pd
import numpy as np

SIZE = 100_000
WARMUP = 5
ITERATIONS = 30

series_data = [float("nan") if i % 5 == 0 else i * 1.0 for i in range(SIZE)]
s = pd.Series(series_data)

df = pd.DataFrame({
    "a": [float("nan") if i % 5 == 0 else i * 0.1 for i in range(SIZE)],
    "b": [float("nan") if i % 7 == 0 else i * 2.0 for i in range(SIZE)],
    "c": [float("nan") if i % 3 == 0 else i % 100 for i in range(SIZE)],
})

for _ in range(WARMUP):
    s.dropna()
    df.dropna()
    df.dropna(how="any")
    df.dropna(how="all")

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    s.dropna()
    df.dropna()
    df.dropna(how="any")
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
mean_ms = total_ms / ITERATIONS
print(json.dumps({"function": "dropna_fn", "mean_ms": round(mean_ms, 3), "iterations": ITERATIONS, "total_ms": round(total_ms, 3)}))

File: benchmarks/tsb/bench_dropna_fn.ts
/**
 * Benchmark: dropnaSeries / dropnaDataFrame — standalone functional dropna.
 * Outputs JSON: {"function": "dropna_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, DataFrame, dropnaSeries, dropnaDataFrame } from "../../src/index.ts";

const SIZE = 100_000;
const WARMUP = 5;
const ITERATIONS = 30;

// ~20% NaN values
const seriesData = Array.from({ length: SIZE }, (_, i) => (i % 5 === 0 ? null : i * 1.0));
const s = new Series({ data: seriesData });

const df = DataFrame.fromColumns({
  a: Array.from({ length: SIZE }, (_, i) => (i % 5 === 0 ? null : i * 0.1)),
  b: Array.from({ length: SIZE }, (_, i) => (i % 7 === 0 ? null : i * 2.0)),
  c: Array.from({ length: SIZE }, (_, i) => (i % 3 === 0 ? null : i % 100)),
});

for (let i = 0; i < WARMUP; i++) {
  dropnaSeries(s);
  dropnaDataFrame(df);
  dropnaDataFrame(df, { how: "any" });
  dropnaDataFrame(df, { how: "all" });
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  dropnaSeries(s);
  dropnaDataFrame(df);
  dropnaDataFrame(df, { how: "any" });
  times.push(performance.now() - t0);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "dropna_fn",
    mean_ms: Math.round(meanMs * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(totalMs * 1000) / 1000,
  }),
);
